alacritty/src/renderer/mod.rs

```diff
@@ -287,14 +295,27 @@ impl GlyphCache {
         let rasterizer = &mut self.rasterizer;
         let metrics = &self.metrics;
         self.cache.entry(glyph_key).or_insert_with(|| {
-            let mut rasterized =
-                rasterizer.get_glyph(glyph_key).unwrap_or_else(|_| Default::default());
+            let mut is_missing = false;
+            let mut rasterized = rasterizer.get_glyph(glyph_key).unwrap_or_else(|err| match err {
+                RasterizerError::MissingGlyph(glyph) => {
+                    is_missing = true;
+                    glyph
+                },
+                _ => Default::default(),
+            });
 
             rasterized.left += i32::from(glyph_offset.x);
             rasterized.top += i32::from(glyph_offset.y);
             rasterized.top -= metrics.descent as i32;
 
-            loader.load_glyph(&rasterized)
+            let mut glyph = loader.load_glyph(&rasterized);
+
+            // Mark glyph as missing.
+            if is_missing {
+                glyph.flags.insert(GlyphFlags::MISSING_GLYPH);
```

Comment:
I don't really have an idea how to do it nicer than that. The whole purpose is to avoid going into crossfont and performing the entire fallback logic on each redraw when zerowidth char is missing in the font, since it could only work if we add missing glyph for that char into the glyph cache, but not drawn on screen.

If you have any idea how to do that better, I'd like to hear, however everything I can think of will require rework of the loading logic entirely, and not sure if it'll be any better than what we have now...

Reply:
I don't like adding this information to the glyph flags at all, logically that doesn't make much sense to me.

Maybe an enum could help out? So we cache either `Glyph { fields }` or `Missing`? The goal is just to get the information that the glyph is missing out of `self.cache.get(key)`, right? Also another more pragmatic solution is to just store the glyph as completely empty, if we don't render anything, there's no need for us to care?

One thing I certainly wouldn't want to do is reserve space in our atlas for missing glyphs, or have duplicate instances of the missing glyph texture in it. So ideally we'd have zero atlas space occupied by missing zerowidth and a single cell for the normal one.